resolwe_bio/kb/management/commands/insert_features.py

```diff
@@ -87,53 +89,89 @@ def add_arguments(self, parser):
+        subtype_choices = list(zip(*Feature.SUBTYPE_CHOICES))[0]
 
-            reader = csv.DictReader(tab_file, delimiter=str('\t'))
-            bar_format = '{desc}{percentage:3.0f}%|{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}]'
+        for tab_file_name, tab_file in decompress(options['file_name']):
+            logger.info(__("Importing features from \"{}\"...", tab_file_name))
+
+            features = []
+            unique_features = set()
+            for row in csv.DictReader(tab_file, delimiter=str('\t')):
+                sub_type = SUBTYPE_MAP.get(row['Gene type'], 'other')
+
+                if row['Type'] not in type_choices:
+                    raise ValidationError("Unknown type: {}".format(row['Type']))
+                if sub_type not in subtype_choices:
+                    raise ValidationError("Unknown subtype: {}".format(sub_type))
 
-            for row in tqdm(reader, total=line_count, bar_format=bar_format):
                 aliases_text = row['Aliases'].strip()
                 aliases = []
                 if aliases_text and aliases_text != '-':
                     aliases = aliases_text.split(',')
 
-                sub_type = SUBTYPE_MAP.get(row['Gene type'], 'other')
-
-                values = {
-                    'species': row['Species'],
-                    'type': row['Type'],
-                    'sub_type': sub_type,
-                    'name': row['Name'],
-                    'full_name': row['Full name'],
-                    'description': row['Description'],
-                    'aliases': aliases,
-                }
-
-                feature, created = Feature.objects.get_or_create(source=row['Source'],
-                                                                 feature_id=row['ID'],
-                                                                 defaults=values)
-                if created:
-                    count_inserted += 1
-                else:
-                    is_update = False
-                    for attr, value in values.items():
-                        if getattr(feature, attr) != value:
-                            setattr(feature, attr, value)
-                            is_update = True
-
-                    if is_update:
-                        feature.save()
-                        count_updated += 1
-                    else:
-                        count_unchanged += 1
-
-        index_builder.push(index=FeatureSearchIndex)
-
-        count_total = count_inserted + count_updated + count_unchanged + count_failed
-        logger.info("Total features: %d. Inserted %d, updated %d, "  # pylint: disable=logging-not-lazy
-                    "unchanged %d, failed %d." %
-                    (count_total, count_inserted, count_updated, count_unchanged, count_failed))
+                # TODO: Raise error in features are not unique
+                if (row['Source'], row['ID']) not in unique_features:
+                    features.append([
```

Comment:
Please add a comment that this is a list of lists and not list of dicts due to performance reasons. And also that any changes here may require the query to be updated.